### DISCUSSION TITLE
Handling of inline require in webpack 5

### DISCUSSION THREAD

**Discussion by daniel-ac-martin** [webpack/webpack#15670] [Q&A] [answered]
I've run into a problem when trying to upgrade from Webpack v4 to v5. The problem is specifically with this line: https://github.com/apollographql/apollo-client/blob/3d09a63f8e859e57fd8065c49e8f0f3787191be0/src/react/ssr/renderToStringWithData.ts#L9

In the imported module the TypeScript code has already been transformed to the following JavaScript:
```js
import { getMarkupFromTree } from "./getDataFromTree.js";
export function renderToStringWithData(component) {
    return getMarkupFromTree({
        tree: component,
        renderFunction: require('react-dom/server').renderToString
    });
}
//# sourceMappingURL=renderToStringWithData.js.map
```

i.e. It's an ESM file but includes an inline `require`.

Under webpack v4 this inline `require` was transformed into a `__webpack_require__` and everything was fine, but under webpack v5 I am left with a raw `require` which fails after packaging as `node_modules/` is no longer available.

Is this expected behaviour?

Can anyone offer me any pointers on how to go about resolving this?

(PS: I'm using Babel with Webpack, please accept my apologies if I've misunderstood the boundaries of responsibility between the two and this is actually a Babel issue.)

Thanks in advance for any help you can provide.

**Comment by daniel-ac-martin** [comment, marked as answer]
Update: Looks like Babel isn't in play for anything in `node_modules` so this should just be about Webpack's handling of inline `require` in `.js` files, in v5 vs v4.

**Reply by daniel-ac-martin** [reply]
Changing the `require` to an inline `import()` causes the code to be transformed to a `__webpack_require__`, albeit wrapped in a promise.

**Reply by daniel-ac-martin** [reply]
Hoisting the inline `import()` to a normal `import` at the top of the file, resolves the problem.

But I have no idea if such a patch would be acceptable to the upstream, as they may have made the `require` inline for a reason.

**Reply by vankop** [reply]
this probably happend because `renderToStringWithData` has type `javascript/esm` https://webpack.js.org/configuration/module/#ruletype

try to explicitly mark this module as `javascript/auto`

**Reply by daniel-ac-martin** [reply]
@vankop: I don't currently have a rule set-up for this file. (And I'd rather not run it through a custom loader if I can help it.)

Do you know what would cause `renderToStringWithData.js` to have a type of `javascript/esm` by default, in v5. (Presumably in v4 it was `javascript/auto`?)

**Reply by daniel-ac-martin** [reply]
@vankop: I can confirm that marking the module as `javascript/auto` does indeed provide a workaround to this problem. Thanks.
(And it looks like I can avoid setting a custom loader as well!)

It would still be good to understand why the difference exists between v4 and v5, and whether anything can be done upstream to avoid the need for this.

Thanks again for the help @vankop!

**Reply by vankop** [reply]
To fully understand whats going on I need a reproducible repo. As I understand from description `require` is not "replaced" by webpack. This could happen only when module type is `javascript/esm` ( could be a different reasons why it is so ). You don't need a loader to force module type. Simple webpack.config for that:
```js
module.exports = {
  module: {
    rules: [
       {
          test: /@apollo/,
          type: "javascript/auto"
       }
    ]
  }
};
```

> It would still be good to understand why the difference exists between v4 and v5, and whether anything can be done upstream to avoid the need for this.

As I remember webpack@4 doesn't have type `javascript/esm`.

**Reply by daniel-ac-martin** [reply]
> As I remember webpack@4 doesn't have type javascript/esm.
Ah, that will be it then. Thanks for the info.
